### ISSUE TITLE
非训练输入尺寸利用onnx->engine推理效果不好的问题

### ISSUE THREAD

**Issue by dingnasty** [sunny2109/SAFMN#63]
作者你好，我现在遇到的问题是训练了一个SAFMN模型，输入尺寸是（1，3，1024，1024），针对pth文件，一张512x512的图直接输入效果很好，若先resize成1024的，再推理效果就很差，当我将pth转成onnx->engine只支持固定输入大小时，我就只能先resize再推理了，效果不尽如人意，有什么好的办法能解决这个问题吗

**Comment by dingnasty** [comment]
这是我修改的train_DF2K_X2.yml

# general settings
# Claculate on 720P
# Params: 0.228M, FLOPs: 51.532G, Activations: 299M
name: SAFMN_FOD_x2
model_type: SRModel
scale: 2
num_gpu: 1  # set num_gpu: 0 for cpu mode
manual_seed: 10

# dataset and data loader settings
datasets:
  train:
    name: FOD_datasets
    type: PairedImageDataset
    dataroot_gt: E:/code/HR/FOD_datasets/train_HR_sub 
    dataroot_lq: E:/code/HR/FOD_datasets/train_LR_bicubic/X2_sub
    filename_tmpl: '{}'
    io_backend:
      type: disk

    gt_size: 128
    use_hflip: true
    use_rot: true

    # data loader
    use_shuffle: true
    num_worker_per_gpu: 0
    batch_size_per_gpu: 24
    dataset_enlarge_ratio: 10
    prefetch_mode: ~

  # we use multiple validation datasets. The SR benchmark datasets can be download from: https://cv.snu.ac.kr/research/EDSR/benchmark.tar
  val:
    name: FOD_datasets_val100
    type: PairedImageDataset
    dataroot_gt: E:/code/HR/FOD_datasets/valid_HR
    dataroot_lq: E:/code/HR/FOD_datasets/valid_LR_bicubic/X2
    filename_tmpl: '{}'
    io_backend:
      type: disk

# network structures
network_g:
  type: SAFMN
  dim: 36
  n_blocks: 8
  ffn_scale: 2.0
  upscaling_factor: 2

# path
path:
  pretrain_network_g: ~
  strict_load_g: false
  resume_state: ~

# training settings
train:
  ema_decay: 0.999
  optim_g:
    type: Adam
    lr: !!float 1e-3
    weight_decay: 0
    betas: [0.9, 0.99]

  scheduler:
    type: CosineAnnealingRestartLR
    periods: [500000]
    restart_weights: [1]
    eta_min: !!float 1e-5

  total_iter: 10000000
  warmup_iter: -1  # no warm up

  # losses
  pixel_opt:
    type: L1Loss
    loss_weight: 1.0
    reduction: mean

  fft_opt:
    type: FFTLoss
    loss_weight: 0.05
    reduction: mean

# validation settings
val:
  val_freq: !!float 5000
  save_img: False
  pbar: False

  metrics:
    psnr:
      type: calculate_psnr
      crop_border: 2
      test_y_channel: False
      better: higher  # the higher, the better. Default: higher
    # ssim:
    #   type: calculate_ssim
    #   crop_border: 2
    #   test_y_channel: False
    #   better: higher  # the higher, the better. Default: higher

# logging settings
logger:
  print_freq: 100
  save_checkpoint_freq: !!float 100000
  use_tb_logger: true
  wandb:
    project: ~
    resume_id: ~

# dist training settings
dist_params:
  backend: nccl
  port: 29500

**Comment by dingnasty** [comment]
这是我修改的train_DF2K_X2.yml改动的部分

# general settings
# Claculate on 720P
# Params: 0.228M, FLOPs: 51.532G, Activations: 299M
name: SAFMN_FOD_x2
model_type: SRModel
scale: 2
num_gpu: 1  # set num_gpu: 0 for cpu mode
manual_seed: 10

# dataset and data loader settings
datasets:
  train:
    name: FOD_datasets
    type: PairedImageDataset
    dataroot_gt: E:/code/HR/FOD_datasets/train_HR_sub 
    dataroot_lq: E:/code/HR/FOD_datasets/train_LR_bicubic/X2_sub
    filename_tmpl: '{}'
    io_backend:
      type: disk

    gt_size: 128
    use_hflip: true
    use_rot: true

    # data loader
    use_shuffle: true
    num_worker_per_gpu: 0
    batch_size_per_gpu: 24
    dataset_enlarge_ratio: 10
    prefetch_mode: ~

  # we use multiple validation datasets. The SR benchmark datasets can be download from: https://cv.snu.ac.kr/research/EDSR/benchmark.tar
  val:
    name: FOD_datasets_val100
    type: PairedImageDataset
    dataroot_gt: E:/code/HR/FOD_datasets/valid_HR
    dataroot_lq: E:/code/HR/FOD_datasets/valid_LR_bicubic/X2
    filename_tmpl: '{}'
    io_backend:
      type: disk

# network structures
network_g:
  type: SAFMN
  dim: 36
  n_blocks: 8
  ffn_scale: 2.0
  upscaling_factor: 2

# path
path:
  pretrain_network_g: ~
  strict_load_g: false
  resume_state: ~

**Comment by dingnasty** [comment]
> `这是我修改的train_DF2K_X2.yml改动的部分

# general settings
# Claculate on 720P
# Params: 0.228M, FLOPs: 51.532G, Activations: 299M
name: SAFMN_FOD_x2
model_type: SRModel
scale: 2
num_gpu: 1  # set num_gpu: 0 for cpu mode
manual_seed: 10

# dataset and data loader settings
datasets:
  train:
    name: FOD_datasets
    type: PairedImageDataset
    dataroot_gt: E:/code/HR/FOD_datasets/train_HR_sub 
    dataroot_lq: E:/code/HR/FOD_datasets/train_LR_bicubic/X2_sub
    filename_tmpl: '{}'
    io_backend:
      type: disk

    gt_size: 128
    use_hflip: true
    use_rot: true

    # data loader
    use_shuffle: true
    num_worker_per_gpu: 0
    batch_size_per_gpu: 24
    dataset_enlarge_ratio: 10
    prefetch_mode: ~

  # we use multiple validation datasets. The SR benchmark datasets can be download from: https://cv.snu.ac.kr/research/EDSR/benchmark.tar
  val:
    name: FOD_datasets_val100
    type: PairedImageDataset
    dataroot_gt: E:/code/HR/FOD_datasets/valid_HR
    dataroot_lq: E:/code/HR/FOD_datasets/valid_LR_bicubic/X2
    filename_tmpl: '{}'
    io_backend:
      type: disk

# network structures
network_g:
  type: SAFMN
  dim: 36
  n_blocks: 8
  ffn_scale: 2.0
  upscaling_factor: 2

# path
path:
  pretrain_network_g: ~
  strict_load_g: false
  resume_state: ~`

**Comment by dingnasty** [comment]
这是我修改之后的yml
# general settings
# Claculate on 720P
# Params: 0.228M, FLOPs: 51.532G, Activations: 299M
name: SAFMN_FOD_x2
model_type: SRModel
scale: 2
num_gpu: 1  # set num_gpu: 0 for cpu mode
manual_seed: 10

# dataset and data loader settings
datasets:
  train:
    name: FOD_datasets
    type: PairedImageDataset
    dataroot_gt: E:/code/HR/FOD_datasets/train_HR_sub 
    dataroot_lq: E:/code/HR/FOD_datasets/train_LR_bicubic/X2_sub
    filename_tmpl: '{}'
    io_backend:
      type: disk

    gt_size: 128
    use_hflip: true
    use_rot: true

    # data loader
    use_shuffle: true
    num_worker_per_gpu: 0
    batch_size_per_gpu: 24
    dataset_enlarge_ratio: 10
    prefetch_mode: ~

  # we use multiple validation datasets. The SR benchmark datasets can be download from: https://cv.snu.ac.kr/research/EDSR/benchmark.tar
  val:
    name: FOD_datasets_val100
    type: PairedImageDataset
    dataroot_gt: E:/code/HR/FOD_datasets/valid_HR
    dataroot_lq: E:/code/HR/FOD_datasets/valid_LR_bicubic/X2
    filename_tmpl: '{}'
    io_backend:
      type: disk

# network structures
network_g:
  type: SAFMN
  dim: 36
  n_blocks: 8
  ffn_scale: 2.0
  upscaling_factor: 2

# path
path:
  pretrain_network_g: ~
  strict_load_g: false
  resume_state: ~

# training settings
train:
  ema_decay: 0.999
  optim_g:
    type: Adam
    lr: !!float 1e-3
    weight_decay: 0
    betas: [0.9, 0.99]

  scheduler:
    type: CosineAnnealingRestartLR
    periods: [500000]
    restart_weights: [1]
    eta_min: !!float 1e-5

  total_iter: 10000000
  warmup_iter: -1  # no warm up

  # losses
  pixel_opt:
    type: L1Loss
    loss_weight: 1.0
    reduction: mean

  fft_opt:
    type: FFTLoss
    loss_weight: 0.05
    reduction: mean

# validation settings
val:
  val_freq: !!float 5000
  save_img: False
  pbar: False

  metrics:
    psnr:
      type: calculate_psnr
      crop_border: 2
      test_y_channel: False
      better: higher  # the higher, the better. Default: higher
    # ssim:
    #   type: calculate_ssim
    #   crop_border: 2
    #   test_y_channel: False
    #   better: higher  # the higher, the better. Default: higher

# logging settings
logger:
  print_freq: 100
  save_checkpoint_freq: !!float 100000
  use_tb_logger: true
  wandb:
    project: ~
    resume_id: ~

# dist training settings
dist_params:
  backend: nccl
  port: 29500

**Comment by sunny2109** [comment]
您可以尝试一下去掉最后的上采样，直接对input做插值，然后再finetune一段时间看看效果。

转onnx不支持任意分辨率的问题，我之前也尝试解决过，暂时没有好的解决方案。
